Add runtime.getContexts documentation

rebloor opened the pull request:
### Description

Documentation in support of [Bug 1875480](https://bugzilla.mozilla.org/show_bug.cgi?id=1875480) Implement runtime.getContexts

### Related issues and pull requests

BCD PR https://github.com/mdn/browser-compat-data/pull/23305

rebloor committed:
Add runtime.getContexts documentation

rebloor committed:
typo

rebloor committed:
Remove auto complete typos

rebloor committed:
Linter code format suggestions

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+
+  - : An object containing criteria to match returned contexts against. Matching contexts must match all specified filters. If the object is empty, all contexts are returned.
+    - `contextIds` {{optional_inline}}
+      - : Array of `string` ID of a context.
```
```suggestion
      - : Array of `string`. ID of contexts to return.
```

Please be consistent in specifying the type; the descriptions are currently a copy-paste of the return values. Rephrase so that it is obvious that it is a filter. For examples, see

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/getSessionRules

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/Tabs/query

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contextualIdentities/query

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/downloads/DownloadQuery

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+- `documentUrl`
+  - : `string` The URL of the document associated with the context or undefined if the context not hosted in a document.
+- `frameId`
+  - : `string` The frame ID for the context or `-1` if the context is not hosted in a frame.
```
```suggestion
  - : `integer` The frame ID for the context or `-1` if the context is not hosted in a frame.
```

(Also in the filter)

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+- `incognito`
+  - : `boolean` Whether the context is hosted in a private browsing tab.
+- `tabId`
+  - : `string` The tab ID for the context or `-1` if the context not hosted in a tab.
```
```suggestion
  - : `integer` The tab ID for the context or `-1` if the context not hosted in a tab.
```

(Also in the filter)

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+- `tabId`
+  - : `string` The tab ID for the context or `-1` if the context not hosted in a tab.
+- `windowId`
+  - : `string` The window ID for the context or `-1` if the context not hosted in a window.
```
```suggestion
  - : `integer` The window ID for the context or `-1` if the context not hosted in a window.
```

(Also for the filter)

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+- `windowId`
+  - : `string` The window ID for the context or `-1` if the context not hosted in a window.
+
+If there are no matching context, fulfilled with an empty array.
```
```suggestion
If there are no matching contexts, fulfilled with an empty array.
```

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
````diff
@@ -0,0 +1,92 @@
+```js
+function gotContextInfo(contexts) {
+  for (const context of contexts) {
+    if ((context.tabId = -1)) {
````
```suggestion
    if (context.tabId == -1) {
```

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+  }
+}
+
+let gettingContextInfo = browser.runtime.getContext([(incognito = true)]);
```
```suggestion
let gettingContextInfo = browser.runtime.getContext({incognito: true});
```

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+    if ((context.tabId = -1)) {
+      console.log("Not hosted in a tab");
+    } else {
+      console.log(`Hosted in tab: ${context.tabId}`);
```
```suggestion
      console.log(`Hosted in tab: ${context.tabId} and frame ${context.frameId} with URL ${context.documentUrl}`);
```

Just the tabId is rarely useful; I have expanded the example to include the frameId and URL.

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+    - `contextTypes` {{optional_inline}}
+      - : Array of `string` The type of extension view. Takes the values `"BACKGROUND"`, `"POPUP"`, `"SIDE_PANEL"`, and `"TAB"`.
+    - `documentIds` {{optional_inline}}
+      - : Array of `string` UUID of a document associated with the context or undefined to match all contexts not hosted in a document.
```
"or undefined" doesn't need to be elaborated, the default assumption is that the property is not filtered on unless specified.

rebloor replied:
```suggestion
      - : Array of `string` UUID of a document associated with the context.
```

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
````diff
@@ -0,0 +1,92 @@
+
+```js-nolint
+let gettingContexts = browser.runtime.getContexts(
+    ContextFilter           // object
````
```suggestion
    filter           // object
```

The variable name should be used, not the "type"

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+
+### Parameters
+
+- `ContextFilter`
```
```suggestion
- `filter`
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+    - `documentUrls` {{optional_inline}}
+      - : Array of `string` The URL of a document associated with the context or undefined for contexts not hosted in a document.
+    - `frameIds` {{optional_inline}}
+      - : Array of `string` The frame ID for the context or `-1` for contexts not hosted in a frame.
```
```suggestion
      - : Array of `integer` The frame ID for the context or `-1` for contexts not hosted in a frame.
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+    - `incognito` {{optional_inline}}
+      - : `boolean` Whether to return only contexts hosted in a private browsing tab.
+    - `tabIds` {{optional_inline}}
+      - : Array of `string` The tab ID for the context or `-1` for contexts not hosted in a tab.
```
```suggestion
      - : Array of `integer` The tab ID for the context or `-1` for contexts not hosted in a tab.
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/runtime/getcontexts/index.md:
```diff
@@ -0,0 +1,92 @@
+    - `tabIds` {{optional_inline}}
+      - : Array of `string` The tab ID for the context or `-1` for contexts not hosted in a tab.
+    - `windowIds` {{optional_inline}}
+      - : Array of `string` The window ID for the context or `-1` for contexts not hosted in a window.
```
```suggestion
      - : Array of `integer` The window ID for the context or `-1` for contexts not hosted in a window.
```

rebloor committed:
Suggestions from review

Co-authored-by: Rob Wu <rob@robwu.nl>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rebloor committed:
Merge branch 'main' into add-runtime.getContexts

rebloor committed:
Apply suggestions from linter

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rebloor committed:
Further feedback updates
- aligned with the current standard for presenting object property data types
- clarified content of filter object
- aligned example description with code update
- fixed various typos and grammatical errors
